cirq-core/cirq/transformers/measurement_transformers.py

```diff
@@ -109,25 +119,34 @@ def defer(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
                 )
                 for indexes, m in gate.confusion_map.items()
             ]
-            cxs = [_mod_add(q, target) for q, target in zip(op.qubits, targets)]
             xs = [ops.X(targets[i]) for i, b in enumerate(gate.full_invert_mask()) if b]
             return cxs + confusions + xs
         elif protocols.is_measurement(op):
             return [defer(op, None) for op in protocols.decompose_once(op)]
         elif op.classical_controls:
-            new_op = op.without_classical_controls()
-            for c in op.classical_controls:
-                if isinstance(c, value.KeyCondition):
-                    if c.key not in measurement_qubits:
-                        raise ValueError(f'Deferred measurement for key={c.key} not found.')
-                    qs = measurement_qubits[c.key]
-                    all_values = itertools.product(*[range(q.dimension) for q in qs])
-                    anything_but_all_zeros = tuple(itertools.islice(all_values, 1, None))
-                    control_values = ops.SumOfProducts(anything_but_all_zeros)
-                    new_op = new_op.controlled_by(*qs, control_values=control_values)
-                else:
-                    raise ValueError('Only KeyConditions are allowed.')
-            return new_op
+            # Convert to a quantum control
+            keys = sorted(set(key for c in op.classical_controls for key in c.keys))
+            for key in keys:
+                if key not in measurement_qubits:
+                    raise ValueError(f'Deferred measurement for key={key} not found.')
+
+            # Try every possible datastore state against the condition, and the ones that work
+            # are the control values for the new op.
+            compatible_datastores = [
+                store
+                for store in _all_possible_datastore_states(keys, measurement_qubits)
+                if all(c.resolve(store) for c in op.classical_controls)
+            ]
```

Comment:
I think it's worth highlighting the exponential complexity in the docstring above.